File: internal/transformers/pull_secret_injector.go
package transformers

import (
	"encoding/json"
	"strings"

	"github.com/abradner/workflow/internal/domain/kubernetes"
	"github.com/abradner/workflow/internal/manifest"
)

// PullSecretInjector synthesizes a registry pull secret (backed by a
// 1Password item) for the app, wires it into ServiceAccounts and the base
// Kustomization, and rewrites overlay deployment image patches to pull
// through the configured registry.
type PullSecretInjector struct {
	RegistryHostname          string
	Registry1PItemID          string
	ExternalSecretsAPIVersion string
	ProjectName               string
}

// Call injects the registry pull secret across the workspace.
func (t PullSecretInjector) Call(ws *manifest.Workspace) *manifest.Workspace {
	dockerConfigBytes, _ := json.Marshal(map[string]any{
		"auths": map[string]any{
			t.RegistryHostname: map[string]any{
				"username": "{{ .username }}",
				"password": "{{ .password }}",
			},
		},
	})

	uniqueSecretName := ws.AppName + "-registry"

	secret := kubernetes.ExternalSecret{
		Name:         uniqueSecretName,
		APIVersion:   t.ExternalSecretsAPIVersion,
		StoreName:    "onepassword-backend",
		TemplateType: "kubernetes.io/dockerconfigjson",
		TemplateData: map[string]any{".dockerconfigjson": string(dockerConfigBytes)},
		DataRefs: []kubernetes.DataRef{
			{SecretKey: "username", Key: t.Registry1PItemID, Property: "username"},
			{SecretKey: "password", Key: t.Registry1PItemID, Property: "password"},
		},
	}

	secretDoc := secret.ToMap()
	spec, _ := secretDoc["spec"].(map[string]any)
	spec["refreshInterval"] = "24h"
	secretDoc["spec"] = spec

	ws.Manifests["base/registry-pull-secret.yaml"] = secretDoc

	for _, path := range ws.SortedPaths() {
		docs := ws.Manifests[path]

		if docsMap, ok := docs.(map[string]any); ok &&
			(strings.Contains(path, "kustomization.yaml") || strings.Contains(path, "kustomization.yml")) {
			if strings.HasPrefix(path, "base/") {
				resources, _ := docsMap["resources"].([]any)
				if !containsString(resources, "registry-pull-secret.yaml") {
					resources = append(resources, "registry-pull-secret.yaml")
				}
				docsMap["resources"] = resources
			}
			continue
		}

		mutated := manifest.MutateYAML(docs, func(d any) any {
			doc, ok := d.(map[string]any)
			if !ok {
				return d
			}
			if doc["kind"] == "ServiceAccount" {
				pullSecrets, _ := doc["imagePullSecrets"].([]any)
				if !containsPullSecretName(pullSecrets, uniqueSecretName) {
					pullSecrets = append(pullSecrets, map[string]any{"name": uniqueSecretName})
				}
				doc["imagePullSecrets"] = pullSecrets
			}
			return doc
		})

		if strings.Contains(path, "deployment.yaml") && strings.HasPrefix(path, "overlay/") {
			if patches, ok := mutated.([]any); ok {
				rewriteImagePatches(patches, t.RegistryHostname, t.ProjectName)
			}
		}

		ws.Manifests[path] = mutated
	}

	return ws
}

func rewriteImagePatches(patches []any, registryHostname, projectName string) {
	for _, patchAny := range patches {
		patch, ok := patchAny.(map[string]any)
		if !ok {
			continue
		}
		patchPath, _ := patch["path"].(string)
		if patch["op"] != "replace" || !strings.Contains(patchPath, "image") {
			continue
		}

		value, _ := patch["value"].(string)
		parts := strings.Split(value, "/")
		imageRepoTag := parts[len(parts)-1]
		patch["value"] = registryHostname + "/" + projectName + "/" + imageRepoTag
	}
}

func containsPullSecretName(secrets []any, name string) bool {
	for _, sAny := range secrets {
		if s, ok := sAny.(map[string]any); ok && s["name"] == name {
			return true
		}
	}
	return false
}

File: ruby-legacy/app/workflow/transformers/pull_secret_injector.rb
# frozen_string_literal: true

require_relative 'base'
require_relative '../../domain/kubernetes/external_secret'
require 'json'

module Workflow
  module Transformers
    # Injects the registry pull secret payload natively into the workspace
    class PullSecretInjector < Base
      def initialize(registry_hostname:, registry_1p_item_id:, external_secrets_api_version:, project_name:)
        @registry_hostname = registry_hostname
        @registry_1p_item_id = registry_1p_item_id
        @external_secrets_api_version = external_secrets_api_version
        @project_name = project_name
      end

      def call(workspace)
        # Synthesize the registry-pull-secret
        dockerconfig = JSON.generate({
          auths: {
            @registry_hostname => {
              username: '{{ .username }}',
              password: '{{ .password }}'
            }
          }
        })

        unique_secret_name = "#{workspace.app_name}-registry"

        secret = Kubernetes::ExternalSecret.new(
          name: unique_secret_name,
          api_version: @external_secrets_api_version,
          store_name: 'onepassword-backend',
          template_type: 'kubernetes.io/dockerconfigjson',
          template_data: { '.dockerconfigjson' => dockerconfig },
          data_refs: [
            { secret_key: 'username', key: @registry_1p_item_id, property: 'username' },
            { secret_key: 'password', key: @registry_1p_item_id, property: 'password' }
          ]
        )

        workspace.manifests['base/registry-pull-secret.yaml'] = secret.to_h.merge(spec: secret.to_h[:spec].merge(refreshInterval: '24h'))

        workspace.manifests.each do |path, docs|
          if docs.is_a?(Hash) && (path.include?('kustomization.yaml') || path.include?('kustomization.yml'))
            # Only add to base kustomizations
            if path.start_with?('base/')
              docs[:resources] ||= []
              docs[:resources] << 'registry-pull-secret.yaml' unless docs[:resources].include?('registry-pull-secret.yaml')
            end
            next
          end

          docs = mutate_yaml(docs) do |doc|
            next doc unless doc.is_a?(Hash)

            # Add image pull secret reference to Service Accounts
            if doc[:kind] == 'ServiceAccount'
              doc[:imagePullSecrets] ||= []
              doc[:imagePullSecrets] << { name: unique_secret_name } unless doc[:imagePullSecrets].any? { |s| s[:name] == unique_secret_name }
            end
            doc
          end

          # Map Registry URIs natively out of overlays
          if path.include?('deployment.yaml') && path.start_with?('overlay/') && docs.is_a?(Array)
            docs.each do |patch|
              if patch.is_a?(Hash) && patch[:op] == 'replace' && patch[:path].to_s.include?('image')
                image_repo_tag = patch[:value].split('/').last
                patch[:value] = "#{@registry_hostname}/#{@project_name}/#{image_repo_tag}"
              end
            end
          end

          workspace.manifests[path] = docs
        end

        workspace
      end
    end
  end
end
